### ISSUE TITLE
Collect: jr:choice-name() fails when editing a saved form that contains a select1 field with using current() in the itemset filter expression.

### ISSUE THREAD

**Issue by meletis** [opendatakit/opendatakit#1180]
This one may be related to #1179, but I can't be sure since the context in #1179 is not defined.

This is form that demonstrates the current issue():

```xml
<?xml version="1.0"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:odk="http://www.opendatakit.org/xforms" xmlns:orx="http://openrosa.org/xforms/" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <h:head>
    <h:title>Test jr:choice-name with current()</h:title>
    <model>
      <itext>
        <translation default="true()" lang="english">
          <text id="static_instance-best-2">
            <value>Orange</value>
          </text>
          <text id="static_instance-best-3">
            <value>Pineapple</value>
          </text>
          <text id="static_instance-best-0">
            <value>Banana</value>
          </text>
          <text id="static_instance-best-1">
            <value>Apple</value>
          </text>
          <text id="static_instance-best-6">
            <value>NFS</value>
          </text>
          <text id="static_instance-best-7">
            <value>FIFA</value>
          </text>
          <text id="static_instance-best-4">
            <value>Age of Kings</value>
          </text>
          <text id="static_instance-best-5">
            <value>Warcarft</value>
          </text>
          <text id="static_instance-types-0">
            <value>Fruits</value>
          </text>
          <text id="static_instance-types-1">
            <value>Games</value>
          </text>
        </translation>
      </itext>
      <instance>
        <testchoicenamewithfilter id="testchoicenamewithfilter" version="1511241215">
          <starttime/>
          <endtime/>
          <deviceid/>
          <subscriberid/>
          <simid/>
          <devicephonenum/>
          <username/>
          <rg_count/>
          <rg jr:template="">
            <type/>
            <labelfromtype/>
            <bestone/>
            <selected/>
            <confirm/>
          </rg>
          <meta>
            <instanceID/>
          </meta>
        </testchoicenamewithfilter>
      </instance>
      <instance id="best">
        <root>
          <item>
            <itextId>static_instance-best-0</itextId>
            <filter>1</filter>
            <name>1</name>
          </item>
          <item>
            <itextId>static_instance-best-1</itextId>
            <filter>1</filter>
            <name>2</name>
          </item>
          <item>
            <itextId>static_instance-best-2</itextId>
            <filter>1</filter>
            <name>3</name>
          </item>
          <item>
            <itextId>static_instance-best-3</itextId>
            <filter>1</filter>
            <name>4</name>
          </item>
          <item>
            <itextId>static_instance-best-4</itextId>
            <filter>2</filter>
            <name>5</name>
          </item>
          <item>
            <itextId>static_instance-best-5</itextId>
            <filter>2</filter>
            <name>6</name>
          </item>
          <item>
            <itextId>static_instance-best-6</itextId>
            <filter>2</filter>
            <name>7</name>
          </item>
          <item>
            <itextId>static_instance-best-7</itextId>
            <filter>2</filter>
            <name>8</name>
          </item>
        </root>
      </instance>
      <instance id="types">
        <root>
          <item>
            <itextId>static_instance-types-0</itextId>
            <name>1</name>
          </item>
          <item>
            <itextId>static_instance-types-1</itextId>
            <name>2</name>
          </item>
        </root>
      </instance>
      <bind jr:preload="timestamp" jr:preloadParams="start" nodeset="/testchoicenamewithfilter/starttime" type="dateTime"/>
      <bind jr:preload="timestamp" jr:preloadParams="end" nodeset="/testchoicenamewithfilter/endtime" type="dateTime"/>
      <bind jr:preload="property" jr:preloadParams="deviceid" nodeset="/testchoicenamewithfilter/deviceid" type="string"/>
      <bind jr:preload="property" jr:preloadParams="subscriberid" nodeset="/testchoicenamewithfilter/subscriberid" type="string"/>
      <bind jr:preload="property" jr:preloadParams="simserial" nodeset="/testchoicenamewithfilter/simid" type="string"/>
      <bind jr:preload="property" jr:preloadParams="phonenumber" nodeset="/testchoicenamewithfilter/devicephonenum" type="string"/>
      <bind jr:preload="property" jr:preloadParams="username" nodeset="/testchoicenamewithfilter/username" type="string"/>
      <bind calculate="2" nodeset="/testchoicenamewithfilter/rg_count" readonly="true()" type="string"/>
      <bind nodeset="/testchoicenamewithfilter/rg/type" required="true()" type="select1"/>
      <bind calculate="jr:choice-name( /testchoicenamewithfilter/rg/type , ' /testchoicenamewithfilter/rg/type ')" nodeset="/testchoicenamewithfilter/rg/labelfromtype" type="string"/>
      <bind nodeset="/testchoicenamewithfilter/rg/bestone" type="select1"/>
      <bind calculate="jr:choice-name( /testchoicenamewithfilter/rg/bestone , ' /testchoicenamewithfilter/rg/bestone ')" nodeset="/testchoicenamewithfilter/rg/selected" type="string"/>
      <bind nodeset="/testchoicenamewithfilter/rg/confirm" original="note" readonly="true()" type="string"/>
      <bind calculate="concat('uuid:', uuid())" nodeset="/testchoicenamewithfilter/meta/instanceID" readonly="true()" type="string"/>
    </model>
  </h:head>
  <h:body>
    <group ref="/testchoicenamewithfilter/rg">
      <label>Repeat group</label>
      <repeat jr:count=" /testchoicenamewithfilter/rg_count " nodeset="/testchoicenamewithfilter/rg">
        <select1 ref="/testchoicenamewithfilter/rg/type">
          <label>What do you want us to ask you about?</label>
          <item>
            <label>Fruits</label>
            <value>1</value>
          </item>
          <item>
            <label>Games</label>
            <value>2</value>
          </item>
        </select1>
        <select1 ref="/testchoicenamewithfilter/rg/bestone">
          <label>Which one do you prefer most?</label>
          <hint><output value=" /testchoicenamewithfilter/rg/labelfromtype "/></hint>
          <itemset nodeset="instance('best')/root/item[filter = current()/type]">
            <value ref="name"/>
            <label ref="jr:itext(itextId)"/>
          </itemset>
        </select1>
        <input ref="/testchoicenamewithfilter/rg/confirm">
          <label>You have selected: <output value=" /testchoicenamewithfilter/rg/selected "/></label></input>
      </repeat>
    </group>
  </h:body>
</h:html>
```

This form has a repeated group with two static instances. If you fill out the form and save and try to edit it later, Collect cannot open it and throws this error:

![image](https://cloud.githubusercontent.com/assets/654163/11364031/31666c90-92a6-11e5-8af6-d7bb744d49c0.png)

The root cause of this is the usage of "current()" in the itemset nodeset filter expression. When the form loads, JR is evaluating jr:choice-name(). For this to happen it tries to populate the dynamic choices using:

> populateDynamicChoices(ItemsetBinding itemset, TreeReference curQRef)

The root cause of the failure is that the "curQRef" is ambiguous ("/testchoicenamewithfilter/rg/bestone") when the form is loading, so the expression "current()/type" resolves to both instances of "type".

**Comment by meletis** [comment]
Here is when the dynamic choices are resolved (in FormDef.populateDynamicChoices(....)):

>      List<TreeReference> matches = itemset.nodesetExpr.evalNodeset(this.getMainInstance(),
>             new EvaluationContext(exprEvalContext, itemset.contextRef.contextualize(curQRef)));

Since curQRef is ambiguous, the "itemset.contextRef" is not contextualized.

**Comment by meletis** [comment]
Interestingly, this behavior was already anticipated, that jr:choice-name() will not work with filtering, as described in the changeset comment below, also noted in the code comment:

https://bitbucket.org/m.sundt/javarosa/commits/73dfb1fcca27b024af3402a45658ba9a355a715a

...since there is a lack of context.

Here is a proposed fix (commented with this issue's url below):

```java
      if (!ec.getFunctionHandlers().containsKey("jr:choice-name")) {
         final FormDef f = this;
         ec.addFunctionHandler(new IFunctionHandler() {
            public String getName() {
               return "jr:choice-name";
            }

            public Object eval(Object[] args, EvaluationContext ec) {
               try {
                  String value = (String) args[0];
                  String questionXpath = (String) args[1];
                  TreeReference ref = RestoreUtils.xfFact.ref(questionXpath);

                  // possible fix for https://github.com/opendatakit/opendatakit/issues/1180
                  if (ref.isAmbiguous()) {
                     ref = ref.contextualize(ec.getContextRef());
                  }

                  QuestionDef q = findQuestionByRef(ref, f);
                  if (q == null
                          || (q.getControlType() != Constants.CONTROL_SELECT_ONE && q
                          .getControlType() != Constants.CONTROL_SELECT_MULTI)) {
                     return "";
                  }

                  // NOTE: this is highly suspect. We have no context against
                  // which to evaluate
                  // a dynamic selection list. This will generally cause that
                  // evaluation to break
                  // if any filtering is done, or, worst case, give unexpected
                  // results.
                  //
                  // We should hook into the existing code (FormEntryPrompt) for
                  // pulling
                  // display text for select choices. however, it's hard,
                  // because we don't really have
                  // any context to work with, and all the situations where that
                  // context would be used
                  // don't make sense for trying to reverse a select value back
                  // to a label in an unrelated
                  // expression

```

We verified that this change fixes the bug in that form but we will need to make sure that this change is safe for every case.

**Comment by meletis** [comment]
Okay, I have created a pull request that includes this fix plus a previous one:

https://bitbucket.org/m.sundt/javarosa/pull-requests/7/two-surveycto-fixes/diff

**Comment by mitchellsundt** [comment]
Incorporated the fix. Thanks!

On Tue, Mar 15, 2016 at 3:39 AM, meletis <notifications@github.com> wrote:

> Okay, I have created a pull request that includes this fix plus a previous
> one:
>
>
> https://bitbucket.org/m.sundt/javarosa/pull-requests/7/two-surveycto-fixes/diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
>
> https://github.com/opendatakit/opendatakit/issues/1180#issuecomment-196760648
>



-- 
Mitch Sundt
Software Engineer
University of Washington
mitchellsundt@gmail.com

**Comment by mitchellsundt** [comment]
Fixed in ODK Collect 1.4.9
